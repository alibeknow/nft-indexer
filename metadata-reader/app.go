package metadata_reader

import (
	"context"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/zeebo/errs"
	"go.uber.org/zap"
	"golang.org/x/sync/errgroup"

	"metadata-reader/database"
	"metadata-reader/private/httpreader"
	"metadata-reader/queue"
	"metadata-reader/server"
)

var (
	requestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "request_per_second_gauge",
			Help: "Number of requests per second",
		},
	)
	messagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "message_per_second_gauge",
			Help: "Number of messages per second",
		},
	)
	instantRequestPerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_request_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)

	instantMessagePerSecond = prometheus.NewGauge(
		prometheus.GaugeOpts{
			Name: "instant_messages_per_second_gauge",
			Help: "Instant number of reqs per second",
		},
	)
)

func init() {
	prometheus.MustRegister(requestPerSecond)
	prometheus.MustRegister(messagePerSecond)
	prometheus.MustRegister(instantRequestPerSecond)
	prometheus.MustRegister(instantMessagePerSecond)
}

type App struct {
	log      *zap.Logger
	db       *database.DB
	srv      *server.Server
	listener *queue.Listener
}

func NewApp(log *zap.Logger, db *database.DB, srv *server.Server, listener *queue.Listener) (*App, error) {
	app := &App{
		log:      log,
		db:       db,
		srv:      srv,
		listener: listener,
	}

	return app, nil
}

func (mr *App) Run(ctx context.Context) error {
	group, ctx := errgroup.WithContext(ctx)

	group.Go(func() error {
		now := time.Now()
		ticker := time.NewTicker(time.Second)
		defer ticker.Stop()

		var (
			reqLast int64
			msgLast int64
		)

		for {
			select {
			case <-ctx.Done():
				return nil
			case <-ticker.C:
				duration := time.Since(now).Seconds()

				reqCount := httpreader.TotalRequestCount()
				requestPerSecond.Set(float64(reqCount) / duration)
				reqDelta := reqCount - reqLast
				reqLast = reqCount
				instantRequestPerSecond.Set(float64(reqDelta))

				msgCount := queue.TotalMessageCount()
				messagePerSecond.Set(float64(msgCount) / duration)
				msgDelta := msgCount - msgLast
				msgLast = msgCount
				instantMessagePerSecond.Set(float64(msgDelta))
			}
		}
	})
	group.Go(func() error {
		return mr.srv.Run(ctx)
	})
	group.Go(func() error {
		return mr.listener.Run(ctx)
	})

	return group.Wait()
}

func (mr *App) Close() error {
	serverErr := mr.srv.Close()
	listenerErr := mr.listener.Close()

	return errs.Combine(serverErr, listenerErr)
}
