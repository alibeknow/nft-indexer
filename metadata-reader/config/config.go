package config

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/spf13/viper"

	"metadata-reader/database"
	"metadata-reader/metadata"
	"metadata-reader/private/logutils"
	"metadata-reader/private/opensearch"
	"metadata-reader/queue"
	"metadata-reader/server"
)

type Config struct {
	Logger   logutils.Config
	Server   server.Config
	Database database.Config
	Queue    queue.Config
	Metadata metadata.Config
	Search   opensearch.Config
}

func LoadConfig(configPaths ...string) (Config, error) {
	v := viper.New()
	v.SetConfigName("config")
	v.SetConfigType("yaml")
	v.SetEnvPrefix("MR")
	v.SetEnvKeyReplacer(strings.NewReplacer(`.`, `_`))
	v.AutomaticEnv()

	v2 := viper.New()
	v2.SetConfigName("config")
	v2.SetConfigType("yaml")
	var buff bytes.Buffer
	m := map[string]interface{}{}
	err := json.NewEncoder(&buff).Encode(Config{})
	if err != nil {
		return Config{}, fmt.Errorf("failed to encode config: %v", err)
	}
	err = json.NewDecoder(&buff).Decode(&m)
	if err != nil {
		return Config{}, fmt.Errorf("failed to decode config: %v", err)
	}
	for key, val := range m {
		v2.Set(key, val)
	}

	for _, path := range configPaths {
		v.AddConfigPath(path)
		v2.AddConfigPath(path)
	}

	if err = v2.SafeWriteConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileAlreadyExistsError); ok {
			log.Println("config.yaml file already exists: skip creating")
		} else {
			return Config{}, fmt.Errorf("failed to write the configuration file: %v", err)
		}
	}

	if err = v.ReadInConfig(); err != nil {
		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
			log.Println("config.yaml ignored")
		} else {
			return Config{}, fmt.Errorf("failed to read the configuration file: %v", err)
		}
	}

	for _, k := range v.AllKeys() {
		value := v.GetString(k)
		v.Set(k, os.ExpandEnv(value))
	}

	var config Config
	err = v.Unmarshal(&config)
	return config, err
}
