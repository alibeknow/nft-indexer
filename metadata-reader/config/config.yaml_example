server:
  address: ":xxx"

database:
  name: "xxx"
  protocol: "mongodb"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "mongodb" for docker
  port: xxx #number
  sslEnable: true/false
  directConnection: true/false

queue:
  protocol: "amqp"
  username: "xxx"
  password: "xxx"
  host: "xxx" #should be "localhost" for local use, and "rabbitmq" for docker
  port: xxx #number
  name: "xxx"

reader:
  workers: xxx #number
  proxyUri: "xxx"
  ipfsGateway: "https://ipfs.io/ipfs/"
  pinataDefaultGateway: "https://gateway.pinata.cloud/ipfs/"

httpreader:
  retries: xxx #number
