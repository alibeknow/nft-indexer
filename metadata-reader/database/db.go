package database

import (
	"context"
	"crypto/tls"
	"crypto/x509"
	"fmt"
	"os"
	"time"

	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"go.mongodb.org/mongo-driver/mongo/readpref"
	"go.mongodb.org/mongo-driver/mongo/writeconcern"
	"go.uber.org/zap"
)

type Config struct {
	Name             string
	Protocol         string
	Username         string
	Password         string
	Host             string
	Port             int
	SSLEnable        bool
	DirectConnection bool
}

func (config *Config) URL() string {
	url := fmt.Sprintf(
		"%s://%s:%s@%s:%d",
		config.Protocol,
		config.Username,
		config.Password,
		config.Host,
		config.Port,
	)

	if config.DirectConnection {
		url = fmt.Sprintf(
			"%s/?connect=direct",
			url,
		)
	}
	return url
}

type DB struct {
	log    *zap.Logger
	client *mongo.Client
	name   string
}

func Open(log *zap.Logger, ctx context.Context, config Config) (*DB, error) {
	opts := options.Client().ApplyURI(config.URL())

	if config.SSLEnable {
		path, _ := os.Getwd()
		pemPath := fmt.Sprintf("%s/../rds-combined-ca-bundle.pem", path)
		rootPEM, err := os.ReadFile(pemPath)
		if err != nil {
			return nil, err
		}

		roots := x509.NewCertPool()
		if ok := roots.AppendCertsFromPEM(rootPEM); !ok {
			return nil, fmt.Errorf("get certs from %s fail!\n", pemPath)
		}
		tlsConfig := &tls.Config{
			RootCAs:            roots,
			InsecureSkipVerify: true,
		}

		opts.SetReadPreference(readpref.Secondary())
		opts.SetWriteConcern(writeconcern.New(writeconcern.WMajority(), writeconcern.J(true), writeconcern.WTimeout(1000)))
		opts.SetTLSConfig(tlsConfig)
	}

	client, err := mongo.NewClient(opts)
	if err != nil {
		return nil, err
	}

	err = client.Connect(ctx)
	if err != nil {
		return nil, err
	}

	pingCtx, pingCancel := context.WithTimeout(ctx, 5*time.Second)
	defer pingCancel()
	err = client.Ping(pingCtx, nil)
	if err != nil {
		return nil, err
	}
	log.Info("Connected to DB", zap.String("url", config.URL()))

	return &DB{
		log:    log,
		client: client,
		name:   config.Name,
	}, nil
}

func (db *DB) Close() error {
	return db.client.Disconnect(context.Background())
}

func (db *DB) Metadata() *MetadataDB {
	coll := db.client.Database(db.name).Collection("metadata")
	return &MetadataDB{
		collection: coll,
	}
}
