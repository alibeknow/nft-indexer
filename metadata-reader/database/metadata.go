package database

import (
	"context"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"

	"metadata-reader/metadata"
)

type MetadataDB struct {
	collection *mongo.Collection
}

func (mDB MetadataDB) InsertOrUpdate(ctx context.Context, m metadata.Metadata) error {
	currentTime := time.Now()

	filter := bson.M{"_id": m.ID}
	update := bson.M{
		"$set": bson.M{
			"metadata":   m.Metadata,
			"type":       m.Type,
			"updated_at": currentTime,
		},
		"$setOnInsert": bson.M{
			"created_at": currentTime,
		},
	}

	opts := options.Update().SetUpsert(true)
	_, err := mDB.collection.UpdateOne(ctx, filter, update, opts)
	return err
}
