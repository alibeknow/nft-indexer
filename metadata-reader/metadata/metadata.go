package metadata

import (
	"context"
	"time"
)

type Metadata struct {
	ID        string     `bson:"_id"`
	Metadata  string     `bson:"metadata"`
	Type      string     `bson:"type"`
	CreatedAt *time.Time `bson:"created_at,omitempty"`
	UpdatedAt *time.Time `bson:"updated_at,omitempty"`
}

type DB interface {
	InsertOrUpdate(ctx context.Context, m Metadata) error
}
