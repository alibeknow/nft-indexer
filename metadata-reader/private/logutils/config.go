package logutils

type Config struct {
	Level string
}
