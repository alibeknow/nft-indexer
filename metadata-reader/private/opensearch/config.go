package opensearch

type Config struct {
	Uri      string
	Shards   int
	Replicas int
}
