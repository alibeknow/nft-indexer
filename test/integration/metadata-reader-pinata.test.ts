process.env.METADATA_REPO_PROVIDER = 'mongo';

import nock from 'nock';
import { ConfigService } from '@nestjs/config';
import { IMetadataReaderConfig, getMetadataReaderConfig } from '@app-metadata-reader/app.config';
import { MetadataReaderService } from '@app-metadata-reader/metadata-reader';
import { DBClass, IDatabaseConfig, getDatabaseConfig } from '@shared/db';
import { MetadataRepoDB } from '@shared/metadata';
import { TokensRepo } from '@shared/tokens';
import { fixturePinataTokens } from './fixtures/tokens.pinata.fixtures';
import { fixtureResponse } from './fixtures/fixture.response';

jest.setTimeout(100000);

describe('[metadata-reader-pinata]', () => {
  let db: DBClass;

  beforeAll(async () => {
    const dbConfig: IDatabaseConfig = getDatabaseConfig();
    db = new DBClass({ ...dbConfig, db: `metadata-reader-pinata-test-${dbConfig.db}` });
    await db.open();

    await db.metadata().deleteMany({});
    await db.tokens().deleteMany({});
  });

  it('should extract metadata from tokens and convert to pinata', async () => {
    await db.tokens().insertMany(fixturePinataTokens);
    const metadataReaderConfig: IMetadataReaderConfig = getMetadataReaderConfig();
    metadataReaderConfig.service.workersCount = 10;
    metadataReaderConfig.service.usePinata = true;
    metadataReaderConfig.service.pinataRetries = 0;
    metadataReaderConfig.service.ipfsGatewayFallback = false;
    metadataReaderConfig.service.defaultGateway = 'https://gateway.pinata.cloud';

    for (const fixtureToken of fixturePinataTokens) {
      nock(fixtureToken.tokenUri).persist().get(/.*/).reply(200, fixtureResponse);
    }

    const configService = new ConfigService<IMetadataReaderConfig>(metadataReaderConfig);
    const r = new MetadataReaderService(
      new TokensRepo(db),
      new MetadataRepoDB(db),
      configService,
    );

    const tokensCount = await r.read(100, 10);
    const tokens = await db.tokens().find({}).toArray();

    // eslint-disable-next-line
    const tokensWithoutCreatedAt = tokens.map(({ createdAt,  ...rest }) => ({ ...rest }));
    const metadataTokens = await db.metadata().countDocuments();

    expect(tokensCount).toEqual(fixturePinataTokens.length);
    expect(metadataTokens).toEqual(fixturePinataTokens.length);
    expect(tokensWithoutCreatedAt).toMatchSnapshot();
  });

  afterEach(async () => {
    await db.tokens().deleteMany({});
    await db.metadata().deleteMany({});
    nock.cleanAll();
  });

  afterAll(async () => {
    await db.close(true);
  });
});
