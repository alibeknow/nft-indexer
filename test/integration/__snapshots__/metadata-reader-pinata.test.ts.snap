// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[metadata-reader-pinata] should extract metadata from tokens and convert to pinata 1`] = `
[
  {
    "_id": "eth:0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f:0x1ecb",
    "block": 14316461,
    "count": 1,
    "from": "0x95d35C8a511F5877Af28B515ab4f0A03B730ae9C",
    "retryCount": 0,
    "tokenMetadataStatus": "available",
    "tokenUri": "https://gateway.pinata.cloud/ipfs/Qmdv7R7C14NqHPLPyQdcCY613w5trQyKHfBsegutRuANF4/7883.json",
  },
  {
    "_id": "eth:0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f:0x04",
    "block": 14316461,
    "count": 1,
    "from": "0x95d35C8a511F5877Af28B515ab4f0A03B730ae9C",
    "retryCount": 0,
    "tokenMetadataStatus": "available",
    "tokenUri": "https://morpheu5.mypinata.cloud/ipfs/QmfLD1MR5uez7xDh28asEuGTUBVBMYQBxvsq6CAcpGuSJb/4",
  },
  {
    "_id": "eth:0x9FB2EEb75754815c5Cc9092Cd53549cEa5dc404f:0x0440",
    "block": 14316462,
    "count": 1,
    "from": "0x95d35C8a511F5877Af28B515ab4f0A03B730ae9C",
    "retryCount": 0,
    "tokenMetadataStatus": "available",
    "tokenUri": "ipfs://QmYGPih59j1BDXPkTysE6To7WQ4uqMAqexQVrunsXjGFj7/1088.json",
  },
  {
    "_id": "eth:0xE1E484251ffFee048A839bc2d68C160BD8C82aBB:0x017e",
    "block": 14316461,
    "count": 1,
    "from": "0x95d35C8a511F5877Af28B515ab4f0A03B730ae9C",
    "retryCount": 0,
    "tokenMetadataStatus": "available",
    "tokenUri": "ipfs://QmdqggiNGU1oxhjpn6t8GUrtCXSSuVmdbTz1FhtRCcnFyy/382.json",
  },
]
`;
